src/blob.jl

```diff
@@ -8,54 +61,99 @@ struct Blob{T}
 
     function Blob{T}(base::Ptr{Nothing}, offset::Int64, limit::Int64) where {T}
         @assert isbitstype(T)
-        new(base, offset, limit)
+        @boundscheck _bounds_check(base, offset, limit, self_size(T), T)
+        new{T}(base, offset, limit)
+    end
+end
+
+@noinline function _bounds_check(
+    base::Ptr{Nothing},
+    offset::Int64,
+    limit::Int64,
+    self_size_T::Int64,
+    @nospecialize(T::DataType))
```

Comment:
Does this `@nospecialize` actually do anything?

Julia doesn't specialize  arguments that look like `t::Type`.  (It does specialize arguments like `::Type{T}) where T`.)
```
julia> foo(base::Int64, t::DataType) = base+sizeof(t)
foo (generic function with 1 method)

julia> Base.specializations(@which foo(3,Int))
Base.MethodSpecializations(MethodInstance for foo(::Int64, ::DataType))

julia> Base.specializations(@which foo(3,UInt8))
Base.MethodSpecializations(MethodInstance for foo(::Int64, ::DataType))

julia> bar(base::Int64, ::Type{T}) where T = base+sizeof(T)
bar (generic function with 1 method)

julia> Base.specializations(@which bar(3,Int))
Base.MethodSpecializations(MethodInstance for bar(::Int64, ::Type{Int64}))

julia> Base.specializations(@which bar(3,UInt8))
Base.MethodSpecializations(svec(MethodInstance for bar(::Int64, ::Type{Int64}), MethodInstance for bar(::Int64, ::Type{UInt8}), nothing, nothing, nothing, nothing, nothing))
```

Reply:
No, good catch. I added it because I was surprised that there is still more code generated by the constructor (I believe). But it didn't change anything for the reason that you point out.